File: Daxelo-Kinrel-Server/src/modules/community/community.service.ts
import {
  Injectable,
  NotFoundException,
  ForbiddenException,
  ConflictException,
  BadRequestException,
  Logger,
} from '@nestjs/common';
import { PrismaService } from '@/common/prisma/prisma.service';
import { CreateCommunityDto } from './dto/create-community.dto';
import { CreateCommentDto } from './dto/create-comment.dto';
import { RsvpDto } from './dto/rsvp.dto';

@Injectable()
export class CommunityService {
  private readonly logger = new Logger(CommunityService.name);

  constructor(private readonly prisma: PrismaService) {}

  // ═══════════════════════════════════════════════════════════════════
  // Community CRUD
  // ═══════════════════════════════════════════════════════════════════

  async listCommunities(options: { type?: string; q?: string; page?: number; limit?: number }) {
    const page = Math.max(1, options.page ?? 1);
    const limit = Math.min(options.limit ?? 20, 100);
    const skip = (page - 1) * limit;

    const where: Record<string, unknown> = {};
    if (options.type) where.type = options.type;
    if (options.q) {
      where.OR = [
        { name: { contains: options.q } },
        { description: { contains: options.q } },
        { gotraName: { contains: options.q } },
        { villageName: { contains: options.q } },
        { surname: { contains: options.q } },
      ];
    }

    const [communities, total] = await Promise.all([
      this.prisma.community.findMany({
        where,
        skip,
        take: limit,
        orderBy: { memberCount: 'desc' },
        include: { _count: { select: { members: true, posts: true, events: true } } },
      }),
      this.prisma.community.count({ where }),
    ]);

    return {
      communities,
      pagination: { page, limit, total, totalPages: Math.ceil(total / limit) },
    };
  }

  async createCommunity(dto: CreateCommunityDto) {
    const user = await this.prisma.user.findUnique({ where: { id: dto.creatorId } });
    if (!user) throw new NotFoundException('User not found');
    if (user.role !== 'admin' && user.role !== 'moderator') {
      throw new ForbiddenException('Insufficient permissions. Moderator or admin role required.');
    }

    const slug = dto.name
      .toLowerCase()
      .replace(/[^a-z0-9\s-]/g, '')
      .replace(/\s+/g, '-')
      .replace(/-+/g, '-')
      .trim();

    const existing = await this.prisma.community.findUnique({ where: { slug } });
    if (existing) throw new ConflictException('A community with this name already exists');

    const community = await this.prisma.community.create({
      data: {
        type: dto.type,
        name: dto.name,
        slug,
        description: dto.description ?? null,
        coverImageUrl: dto.coverImageUrl ?? null,
        iconUrl: dto.iconUrl ?? null,
        isPrivate: dto.isPrivate ?? false,
        gotraName: dto.gotraName ?? null,
        villageName: dto.villageName ?? null,
        surname: dto.surname ?? null,
        region: dto.region ?? null,
        memberCount: 1,
      },
    });

    await this.prisma.communityMember.create({
      data: { communityId: community.id, userId: dto.creatorId, role: 'admin', joinedVia: 'creation' },
    });

    return { community };
  }

  async getCommunity(communityId: string) {
    const community = await this.prisma.community.findUnique({
      where: { id: communityId },
      include: {
        members: { take: 10, orderBy: { joinedAt: 'desc' }, include: { user: { select: { id: true, name: true, email: true } } } },
        rules: { orderBy: { sortOrder: 'asc' } },
        _count: { select: { members: true, posts: true, events: true } },
      },
    });

    if (!community) throw new NotFoundException('Community not found');

    const recentPosts = await this.prisma.communityPost.count({
      where: { communityId, createdAt: { gte: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000) } },
    });

    return {
      community,
      stats: {
        totalMembers: community._count.members,
        totalPosts: community._count.posts,
        totalEvents: community._count.events,
        recentPostsThisWeek: recentPosts,
      },
    };
  }

  async updateCommunity(communityId: string, body: Record<string, unknown>) {
    const { userId, name, description, coverImageUrl, iconUrl, isPrivate, isVerified, region } = body;

    if (!userId) throw new BadRequestException('userId is required');

    const membership = await this.prisma.communityMember.findFirst({
      where: { communityId, userId: userId as string, role: 'admin' },
    });
    if (!membership) throw new ForbiddenException('Only community admins can update the community');

    const updateData: Record<string, unknown> = {};
    if (name !== undefined) updateData.name = name;
    if (description !== undefined) updateData.description = description;
    if (coverImageUrl !== undefined) updateData.coverImageUrl = coverImageUrl;
    if (iconUrl !== undefined) updateData.iconUrl = iconUrl;
    if (isPrivate !== undefined) updateData.isPrivate = isPrivate;
    if (region !== undefined) updateData.region = region;

    if (isVerified !== undefined) {
      const user = await this.prisma.user.findUnique({ where: { id: userId as string } });
      if (user?.role === 'admin') updateData.isVerified = isVerified;
    }

    const community = await this.prisma.community.update({
      where: { id: communityId },
      data: updateData,
    });

    return { community };
  }

  async deleteCommunity(communityId: string, userId: string) {
    if (!userId) throw new BadRequestException('userId is required');

    const membership = await this.prisma.communityMember.findFirst({
      where: { communityId, userId, role: 'admin' },
    });
    if (!membership) throw new ForbiddenException('Only community admins can delete the community');

    await this.prisma.community.delete({ where: { id: communityId } });
    return { success: true };
  }

  async joinOrLeave(communityId: string, body: { userId: string; action: string }) {
    const { userId, action } = body;
    if (!userId) throw new BadRequestException('userId is required');

    const community = await this.prisma.community.findUnique({ where: { id: communityId } });
    if (!community) throw new NotFoundException('Community not found');

    if (action === 'leave') {
      const membership = await this.prisma.communityMember.findUnique({
        where: { communityId_userId: { communityId, userId } },
      });
      if (!membership) throw new BadRequestException('Not a member of this community');

      if (membership.role === 'admin') {
        const adminCount = await this.prisma.communityMember.count({ where: { communityId, role: 'admin' } });
        if (adminCount <= 1) throw new BadRequestException('Cannot leave as the last admin. Transfer admin role first.');
      }

      await this.prisma.communityMember.delete({ where: { communityId_userId: { communityId, userId } } });
      await this.prisma.community.update({ where: { id: communityId }, data: { memberCount: { decrement: 1 } } });

      return { success: true, action: 'left' };
    }

    // JOIN
    const existing = await this.prisma.communityMember.findUnique({
      where: { communityId_userId: { communityId, userId } },
    });
    if (existing) {
      if (existing.role === 'banned') throw new ForbiddenException('You have been banned from this community');
      throw new BadRequestException('Already a member of this community');
    }

    let joinVia = 'search';
    if (community.isPrivate) joinVia = 'invitation';

    // Auto-detect gotra/village/surname match
    if (community.gotraName || community.villageName || community.surname) {
      const userFamilies = await this.prisma.familyMember.findMany({
        where: { userId },
        include: { family: true },
      });
      for (const fm of userFamilies) {
        if (community.gotraName && fm.family.gotra === community.gotraName) { joinVia = 'auto_gotra'; break; }
        if (community.villageName && fm.family.originVillage === community.villageName) { joinVia = 'auto_gotra'; break; }
      }
    }

    await this.prisma.communityMember.create({
      data: { communityId, userId, role: 'member', joinedVia: joinVia },
    });
    await this.prisma.community.update({ where: { id: communityId }, data: { memberCount: { increment: 1 } } });

    return { success: true, action: 'joined', joinVia };
  }

  // ═══════════════════════════════════════════════════════════════════
  // Feed
  // ═══════════════════════════════════════════════════════════════════

  async getFeed(options: { userId: string; cursor?: string; limit?: number; types?: string }) {
    const { userId } = options;
    const limit = Math.min(options.limit ?? 30, 100);

    const familyMemberships = await this.prisma.familyMember.findMany({ where: { userId }, select: { familyId: true } });
    const familyIds = familyMemberships.map((m) => m.familyId);

    const communityMemberships = await this.prisma.communityMember.findMany({ where: { userId }, select: { communityId: true } });
    const communityIds = communityMemberships.map((m) => m.communityId);

    // Get community posts
    const postsWhere: Record<string, unknown>[] = [];
    if (communityIds.length > 0) postsWhere.push({ communityId: { in: communityIds } });
    if (familyIds.length > 0) postsWhere.push({ familyId: { in: familyIds } });
    postsWhere.push({ visibility: 'public' });

    const posts = await this.prisma.communityPost.findMany({
      where: {
        OR: postsWhere,
        isHidden: false,
        ...(options.cursor ? { createdAt: { lt: new Date(options.cursor) } } : {}),
      },
      take: limit,
      orderBy: { createdAt: 'desc' },
      include: {
        author: { select: { id: true, name: true } },
        _count: { select: { reactions: true, comments: true } },
      },
    });

    const items = posts.map((post) => ({
      id: post.id,
      type: post.type,
      title: post.title ?? '',
      body: post.body,
      authorId: post.authorId,
      authorName: post.author.name ?? 'Unknown',
      familyId: post.familyId ?? undefined,
      communityId: post.communityId ?? undefined,
      mediaUrls: JSON.parse(post.mediaUrls || '[]'),
      createdAt: post.createdAt,
      reactionCount: post._count.reactions,
      commentCount: post._count.comments,
    }));

    // Get upcoming events
    const events = await this.prisma.communityEvent.findMany({
      where: {
        OR: [
          { familyId: { in: familyIds } },
          { communityId: { in: communityIds } },
          { visibility: 'public' },
        ],
        startDate: { gte: new Date() },
        isCancelled: false,
      },
      take: 10,
      orderBy: { startDate: 'asc' },
      include: { _count: { select: { rsvps: true } } },
    });

    for (const event of events) {
      items.push({
        id: `event-${event.id}`,
        type: 'event',
        title: event.title,
        body: event.description ?? '',
        authorId: event.creatorId,
        authorName: 'Event',
        familyId: event.familyId ?? undefined,
        communityId: event.communityId ?? undefined,
        mediaUrls: [],
        createdAt: event.createdAt,
        reactionCount: 0,
        commentCount: event._count.rsvps,
      });
    }

    // Sort by createdAt descending
    items.sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());

    const paginated = items.slice(0, limit);
    const nextCursor = paginated.length > 0 ? paginated[paginated.length - 1].createdAt.toISOString() : null;

    return {
      items: paginated,
      pagination: { nextCursor, hasMore: items.length > limit, count: paginated.length },
    };
  }

  // ═══════════════════════════════════════════════════════════════════
  // Comments
  // ═══════════════════════════════════════════════════════════════════

  async getComments(postId: string, options: { parentId?: string; page?: number; limit?: number }) {
    const page = Math.max(1, options.page ?? 1);
    const limit = Math.min(options.limit ?? 50, 100);
    const skip = (page - 1) * limit;

    const where: Record<string, unknown> = { postId, isHidden: false };
    if (options.parentId) {
      where.parentId = options.parentId;
    } else {
      where.parentId = null;
    }

    const [comments, total] = await Promise.all([
      this.prisma.comment.findMany({
        where,
        skip,
        take: limit,
        orderBy: { createdAt: 'asc' },
        include: { author: { select: { id: true, name: true } } },
      }),
      this.prisma.comment.count({ where }),
    ]);

    return {
      comments,
      pagination: { page, limit, total, totalPages: Math.ceil(total / limit) },
    };
  }

  async createComment(postId: string, dto: CreateCommentDto) {
    const post = await this.prisma.communityPost.findUnique({ where: { id: postId } });
    if (!post) throw new NotFoundException('Post not found');
    if (post.isLocked) throw new ForbiddenException('This post is locked for new comments');

    if (dto.parentId) {
      const parentComment = await this.prisma.comment.findUnique({ where: { id: dto.parentId } });
      if (!parentComment || parentComment.postId !== postId) throw new NotFoundException('Parent comment not found');
      if (parentComment.parentId) throw new BadRequestException('Only one level of comment nesting is supported');
    }

    const comment = await this.prisma.comment.create({
      data: {
        postId,
        authorId: dto.authorId,
        parentId: dto.parentId ?? null,
        body: dto.body.trim(),
      },
      include: { author: { select: { id: true, name: true } } },
    });

    return { comment };
  }

  // ═══════════════════════════════════════════════════════════════════
  // Reactions
  // ═══════════════════════════════════════════════════════════════════

  async getReactions(postId: string, userId?: string) {
    const reactions = await this.prisma.reaction.findMany({
      where: { postId },
      include: { user: { select: { id: true, name: true } } },
    });

    const aggregated: Record<string, { count: number; users: Array<{ id: string; name: string | null }> }> = {};
    for (const reaction of reactions) {
      if (!aggregated[reaction.emoji]) aggregated[reaction.emoji] = { count: 0, users: [] };
      aggregated[reaction.emoji].count++;
      aggregated[reaction.emoji].users.push({ id: reaction.user.id, name: reaction.user.name });
    }

    let userReactions: string[] = [];
    if (userId) {
      const myReactions = await this.prisma.reaction.findMany({
        where: { postId, userId },
        select: { emoji: true },
      });
      userReactions = myReactions.map((r) => r.emoji);
    }

    return { aggregated, userReactions, totalReactions: reactions.length };
  }

  async toggleReaction(postId: string, body: { userId: string; emoji: string }) {
    const post = await this.prisma.communityPost.findUnique({ where: { id: postId } });
    if (!post) throw new NotFoundException('Post not found');

    const existing = await this.prisma.reaction.findUnique({
      where: { postId_userId_emoji: { postId, userId: body.userId, emoji: body.emoji } },
    });

    if (existing) {
      await this.prisma.reaction.delete({ where: { id: existing.id } });
      return { action: 'removed', emoji: body.emoji };
    } else {
      await this.prisma.reaction.create({ data: { postId, userId: body.userId, emoji: body.emoji } });
      return { action: 'added', emoji: body.emoji };
    }
  }

  // ═══════════════════════════════════════════════════════════════════
  // Event RSVP
  // ═══════════════════════════════════════════════════════════════════

  async rsvp(eventId: string, dto: RsvpDto) {
    const event = await this.prisma.communityEvent.findUnique({ where: { id: eventId } });
    if (!event) throw new NotFoundException('Event not found');
    if (event.isCancelled) throw new BadRequestException('Cannot RSVP to a cancelled event');

    const existingRSVP = await this.prisma.eventRSVP.findUnique({
      where: { eventId_userId: { eventId, userId: dto.userId } },
    });

    let rsvp;
    if (existingRSVP) {
      rsvp = await this.prisma.eventRSVP.update({
        where: { id: existingRSVP.id },
        data: {
          status: dto.status,
          plusOne: dto.plusOne ?? existingRSVP.plusOne,
          note: dto.note ?? existingRSVP.note,
          respondedAt: new Date(),
        },
      });
    } else {
      rsvp = await this.prisma.eventRSVP.create({
        data: {
          eventId,
          userId: dto.userId,
          status: dto.status,
          plusOne: dto.plusOne ?? false,
          note: dto.note ?? null,
          respondedAt: new Date(),
        },
      });
    }

    const rsvpSummary = await this.prisma.eventRSVP.groupBy({
      by: ['status'],
      where: { eventId },
      _count: { status: true },
    });

    const summary: Record<string, number> = { pending: 0, attending: 0, maybe: 0, declined: 0 };
    for (const row of rsvpSummary) {
      summary[row.status] = row._count.status;
    }

    return { rsvp, summary };
  }
}

File: Daxelo-Kinrel-App/lib/features/community/services/community_service.dart
// lib/features/community/services/community_service.dart
//
// DAXELO KINREL — Community API Service
//
// Calls the NestJS community backend endpoints.
// Uses the shared dioProvider for authenticated HTTP requests.

import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../../core/networking/dio_client.dart';

// ═══════════════════════════════════════════════════════════════════════
// API Response Models
// ═══════════════════════════════════════════════════════════════════════

class CommunityModel {
  const CommunityModel({
    required this.id,
    required this.name,
    required this.type,
    this.description,
    this.avatarUrl,
    this.coverUrl,
    this.isPublic = true,
    this.memberCount = 0,
    this.postCount = 0,
    this.gotraName,
    this.villageName,
    this.surname,
    this.region,
    this.rules = const [],
    this.myRole,
    required this.createdAt,
  });

  final String id;
  final String name;
  final String type; // gotra, village, surname, custom
  final String? description;
  final String? avatarUrl;
  final String? coverUrl;
  final bool isPublic;
  final int memberCount;
  final int postCount;
  final String? gotraName;
  final String? villageName;
  final String? surname;
  final String? region;
  final List<Map<String, dynamic>> rules;
  final String? myRole; // admin, moderator, member, null (not joined)
  final DateTime createdAt;

  bool get isJoined => myRole != null;
  bool get isAdmin => myRole == 'admin';
  bool get isModerator => myRole == 'moderator' || myRole == 'admin';

  factory CommunityModel.fromJson(Map<String, dynamic> json) {
    return CommunityModel(
      id: json['id'] as String,
      name: json['name'] as String,
      type: json['type'] as String? ?? 'custom',
      description: json['description'] as String?,
      avatarUrl: json['avatarUrl'] as String?,
      coverUrl: json['coverUrl'] as String?,
      isPublic: json['isPublic'] as bool? ?? true,
      memberCount: json['memberCount'] as int? ?? 0,
      postCount: json['postCount'] as int? ?? 0,
      gotraName: json['gotraName'] as String?,
      villageName: json['villageName'] as String?,
      surname: json['surname'] as String?,
      region: json['region'] as String?,
      rules: (json['rules'] as List<dynamic>?)
              ?.map((e) => e as Map<String, dynamic>)
              .toList() ??
          [],
      myRole: json['myRole'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

class CommunityPostModel {
  const CommunityPostModel({
    required this.id,
    required this.communityId,
    required this.authorId,
    required this.type,
    required this.body,
    this.title,
    this.mediaUrls = const [],
    this.visibility = 'members_only',
    this.isPinned = false,
    this.isLocked = false,
    this.createdAt,
  });

  final String id;
  final String communityId;
  final String authorId;
  final String type; // discussion, announcement, poll, media
  final String body;
  final String? title;
  final List<String> mediaUrls;
  final String visibility;
  final bool isPinned;
  final bool isLocked;
  final DateTime? createdAt;

  factory CommunityPostModel.fromJson(Map<String, dynamic> json) {
    return CommunityPostModel(
      id: json['id'] as String,
      communityId: json['communityId'] as String,
      authorId: json['authorId'] as String? ?? '',
      type: json['type'] as String? ?? 'discussion',
      body: json['body'] as String,
      title: json['title'] as String?,
      mediaUrls: (json['mediaUrls'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      visibility: json['visibility'] as String? ?? 'members_only',
      isPinned: json['isPinned'] as bool? ?? false,
      isLocked: json['isLocked'] as bool? ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
    );
  }
}

class CommunityEventModel {
  const CommunityEventModel({
    required this.id,
    required this.communityId,
    required this.title,
    this.description,
    this.eventDate,
    this.location,
    this.myRsvp,
    this.attendeeCount = 0,
    required this.createdAt,
  });

  final String id;
  final String communityId;
  final String title;
  final String? description;
  final DateTime? eventDate;
  final String? location;
  final String? myRsvp; // attending, maybe, declined
  final int attendeeCount;
  final DateTime createdAt;

  factory CommunityEventModel.fromJson(Map<String, dynamic> json) {
    return CommunityEventModel(
      id: json['id'] as String,
      communityId: json['communityId'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      eventDate: json['eventDate'] != null
          ? DateTime.parse(json['eventDate'] as String)
          : null,
      location: json['location'] as String?,
      myRsvp: json['myRsvp'] as String?,
      attendeeCount: json['attendeeCount'] as int? ?? 0,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

// ═══════════════════════════════════════════════════════════════════════
// CommunityService — API calls
// ═══════════════════════════════════════════════════════════════════════

class CommunityService {
  CommunityService(this._ref);
  final Ref _ref;

  static const _basePath = '/v1/communities';

  /// Search/browse communities.
  Future<List<CommunityModel>> searchCommunities({
    String? type,
    String? search,
    int limit = 20,
    int page = 1,
  }) async {
    final dio = _ref.read(dioProvider);
    final query = <String, dynamic>{
      'limit': limit,
      'page': page,
      if (type != null) 'type': type,
      if (search != null) 'search': search,
    };
    final response = await dio.get(_basePath, queryParameters: query);
    final data = response.data;
    // Handle both paginated and array responses
    final list = data is List ? data : (data['items'] as List<dynamic>? ?? []);
    return list.map((e) => CommunityModel.fromJson(e as Map<String, dynamic>)).toList();
  }

  /// Create a new community.
  Future<CommunityModel> createCommunity({
    required String name,
    required String type,
    String? description,
    bool isPublic = true,
    String? gotraName,
    String? villageName,
    String? surname,
    String? region,
  }) async {
    final dio = _ref.read(dioProvider);
    final response = await dio.post(_basePath, data: {
      'name': name,
      'type': type,
      if (description != null) 'description': description,
      'isPublic': isPublic,
      if (gotraName != null) 'gotraName': gotraName,
      if (villageName != null) 'villageName': villageName,
      if (surname != null) 'surname': surname,
      if (region != null) 'region': region,
    });
    return CommunityModel.fromJson(response.data as Map<String, dynamic>);
  }

  /// Get community detail.
  Future<CommunityModel> getCommunity(String communityId) async {
    final dio = _ref.read(dioProvider);
    final response = await dio.get('$_basePath/$communityId');
    return CommunityModel.fromJson(response.data as Map<String, dynamic>);
  }

  /// Join a community.
  Future<Map<String, dynamic>> joinCommunity(String communityId) async {
    final dio = _ref.read(dioProvider);
    final response = await dio.post('$_basePath/$communityId/join');
    return response.data as Map<String, dynamic>;
  }

  /// Leave a community.
  Future<void> leaveCommunity(String communityId) async {
    final dio = _ref.read(dioProvider);
    await dio.post('$_basePath/$communityId/leave');
  }

  /// Get community posts.
  Future<List<CommunityPostModel>> getPosts(String communityId, {int limit = 20, int page = 1}) async {
    final dio = _ref.read(dioProvider);
    final response = await dio.get('$_basePath/$communityId/posts', queryParameters: {
      'limit': limit,
      'page': page,
    });
    final data = response.data;
    final list = data is List ? data : (data['items'] as List<dynamic>? ?? []);
    return list.map((e) => CommunityPostModel.fromJson(e as Map<String, dynamic>)).toList();
  }

  /// Create a community post.
  Future<CommunityPostModel> createPost(String communityId, {
    required String type,
    required String body,
    String? title,
    List<String>? mediaUrls,
    String visibility = 'members_only',
  }) async {
    final dio = _ref.read(dioProvider);
    final response = await dio.post('$_basePath/$communityId/posts', data: {
      'type': type,
      'body': body,
      if (title != null) 'title': title,
      if (mediaUrls != null) 'mediaUrls': mediaUrls,
      'visibility': visibility,
    });
    return CommunityPostModel.fromJson(response.data as Map<String, dynamic>);
  }

  /// Get community events.
  Future<List<CommunityEventModel>> getEvents(String communityId, {String? filter, int limit = 20, int page = 1}) async {
    final dio = _ref.read(dioProvider);
    final query = <String, dynamic>{
      'limit': limit,
      'page': page,
      if (filter != null) 'filter': filter,
    };
    final response = await dio.get('$_basePath/$communityId/events', queryParameters: query);
    final data = response.data;
    final list = data is List ? data : (data['items'] as List<dynamic>? ?? []);
    return list.map((e) => CommunityEventModel.fromJson(e as Map<String, dynamic>)).toList();
  }

  /// Create a community event.
  Future<CommunityEventModel> createEvent(String communityId, {
    required String title,
    String? description,
    DateTime? eventDate,
    String? location,
  }) async {
    final dio = _ref.read(dioProvider);
    final response = await dio.post('$_basePath/$communityId/events', data: {
      'title': title,
      if (description != null) 'description': description,
      if (eventDate != null) 'eventDate': eventDate.toIso8601String(),
      if (location != null) 'location': location,
    });
    return CommunityEventModel.fromJson(response.data as Map<String, dynamic>);
  }

  /// RSVP to an event.
  Future<Map<String, dynamic>> rsvpEvent(String communityId, String eventId, {required String status}) async {
    final dio = _ref.read(dioProvider);
    final response = await dio.post('$_basePath/$communityId/events/$eventId/rsvp', data: {
      'status': status,
    });
    return response.data as Map<String, dynamic>;
  }

  /// Get pending join requests (admin only).
  Future<List<Map<String, dynamic>>> getPendingRequests(String communityId) async {
    final dio = _ref.read(dioProvider);
    final response = await dio.get('$_basePath/$communityId/members/requests');
    return (response.data as List<dynamic>).map((e) => e as Map<String, dynamic>).toList();
  }

  /// Approve a join request.
  Future<void> approveRequest(String communityId, String userId) async {
    final dio = _ref.read(dioProvider);
    await dio.post('$_basePath/$communityId/members/$userId/approve');
  }

  /// Reject a join request.
  Future<void> rejectRequest(String communityId, String userId) async {
    final dio = _ref.read(dioProvider);
    await dio.post('$_basePath/$communityId/members/$userId/reject');
  }
}

// ═══════════════════════════════════════════════════════════════════════
// Provider
// ═══════════════════════════════════════════════════════════════════════

final communityServiceProvider = Provider<CommunityService>((ref) {
  return CommunityService(ref);
});
